src/EFCore.PG/Storage/Internal/Mapping/NpgsqlReadOnlyHstoreTypeMapping.cs

```diff
@@ -0,0 +1,46 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.Mapping
+{
+    /// <summary>
+    /// The type mapping for the PostgreSQL hstore type to immutable .NET Dictionaries.
+    /// </summary>
+    /// <remarks>
+    /// See: https://www.postgresql.org/docs/current/static/hstore.html
+    /// </remarks>
+    public class NpgsqlReadOnlyHstoreTypeMapping : NpgsqlHstoreTypeMapping
+    {
+        static readonly HstoreComparer ComparerInstance = new HstoreComparer();
+
+        /// <summary>
+        /// Constructs an instance of the <see cref="NpgsqlReadOnlyHstoreTypeMapping"/> class.
+        /// </summary>
+        public NpgsqlReadOnlyHstoreTypeMapping()
+            : base(
+                new RelationalTypeMappingParameters(
+                    new CoreTypeMappingParameters(typeof(IReadOnlyDictionary<string, string>), null, ComparerInstance),
+                    "hstore"
+                )) {}
+
+        protected NpgsqlReadOnlyHstoreTypeMapping(RelationalTypeMappingParameters parameters)
+            : base(parameters) {}
+
+        protected override RelationalTypeMapping Clone(RelationalTypeMappingParameters parameters)
+            => new NpgsqlReadOnlyHstoreTypeMapping(parameters);
+
+        class HstoreComparer : ValueComparer<IReadOnlyDictionary<string, string>>
```

Comment:
/cc @ajcvickers for a really nice use of a .NET immutable type for version tracking purposes. tl;dr PG has a type called "hstore", which corresponds to .NET Dictionary<string, string?>. This PR also allows you to map hstore to .NET Immutable<string, string?>`, which (hopefully efficiently) creates copies on updates. This means we can do simple reference comparisons, and skip actually snapshotting altogether.

Could be interesting to think about for other types (e.g. ImmutableArray<byte> for mapping blobs/varbinary??).